src/globus_cli/commands/flows/create.py

```diff
@@ -0,0 +1,196 @@
+    """,
+)
+@click.option(
+    "--subtitle",
+    type=str,
+    help="A concise summary of the flow's purpose.",
+)
+@click.option(
+    "--description",
+    type=str,
+    help="A detailed description of the flow's purpose.",
+)
+@click.option(
+    "--administrator",
+    "administrators",
+    type=str,
+    multiple=True,
+    help="""
+        A principal that may perform administrative operations
+        on the flow (e.g., update, delete).
+
+        This option can be specified multiple times
+        to create a list of flow administrators.
+    """,
+)
+@click.option(
+    "--starter",
+    "starters",
+    type=str,
+    multiple=True,
+    help="""
+        A principal that may run the flow.
+
+        Use "all_authenticated_users" to make the flow visible to everyone.
+
+        This option can be specified multiple times
+        to create a list of flow starters.
+    """,
+)
+@click.option(
+    "--viewer",
+    "viewers",
+    type=str,
+    multiple=True,
+    help="""
+        A principal that may view the flow.
+
+        Use "public" to allow any authenticated user to run the flow.
+
+        This option can be specified multiple times
+        to create a list of flow viewers.
+    """,
+)
+@click.option(
+    "--keyword",
+    "keywords",
+    type=str,
+    multiple=True,
+    help="""
+        A term used to help discover this flow when
+        browsing and searching.
+
+        This option can be specified multiple times
+        to create a list of keywords.
+    """,
+)
+@LoginManager.requires_login("flows")
+def create_command(
+    login_manager: LoginManager,
+    title: str,
+    definition: dict,
+    input_schema: dict | None | t.Any,
+    subtitle: str | None,
+    description: str | None,
+    administrators: tuple[str],
+    starters: tuple[str],
+    viewers: tuple[str],
+    keywords: tuple[str],
+) -> None:
+    """
+    Create a new flow.
+
+    TITLE is the name of the flow.
+
+    DEFINITION is the JSON document that defines the flow's instructions.
+    The definition document may be specified inline,
+    or it may be a path to a JSON file, prefixed with "file:".
+
+        Example: Inline JSON:
+
+        \b
+            globus flows create 'My Cool Flow' \\
+            '{{"StartAt": "a", "States": {{"a": {{"Type": "Pass", "End": true}}}}}}'
+
+        Example: Path to JSON file:
+
+        \b
+            globus flows create 'My Other Flow' file:definition.json
+    """
+
+    if input_schema is None:
+        input_schema = {}
+
+    flows_client = login_manager.get_flows_client()
+    auth_client = login_manager.get_auth_client()
+
+    res = flows_client.create_flow(
+        title=title,
+        definition=definition,
+        input_schema=input_schema,
+        subtitle=subtitle,
+        description=description,
+        flow_viewers=list(viewers),
+        flow_starters=list(starters),
+        flow_administrators=list(administrators),
```

Comment:
The current behavior here will be to pass these inputs verbatim, which means that users won't be able to pass usernames or unadorned UUIDs (which we could then decorate as URNs).

I could see an argument against accepting bare UUIDs, but I'd like us to accept usernames. They're much more user-friendly.

What about doing the following to each of these?
```python
new = []
for p in principal_list:
    if p.startswith("urn:"):
        new.append(p)
        continue
    resolved = auth_client.maybe_lookup_identity_id(p)
    if resolved:
        new.append(f"urn:globus:auth:identity:id:{resolved}"
    else:
        raise click.UsageError("bad, no such user")
```

Note that the way `maybe_lookup_identity_id` is written, it will treat a UUID with a no-op. We could adjust that for this case, or explicitly check for (and reject) UUIDs.

Reply:
I'm pretty sure I am a 'no' on bare UUIDs, but I think—though I was originally leaning against it— I've come around on attempting to resolve usernames. I'll update to address.

Reply:
As I am starting to work on this, a couple things that I'm realizing:

a) this kind of contradicts the prior conversation we had as a group about wanting to have the validation of these values happen at the API-level so that it's controlled by the service. I'd originally had written a principal ParamType that did stricter validation here of the values along with nice user messaging, but I felt like there was some debate over whether we shouldn't just be passing more or less anything the user inputs here and let the CLI sort it out, so I put that on ice in order to move this forward.

b) this gets uglier when we start to add handling for the special principals

I'm up for talking through how we want to address, but would sort of like to stet this for the scope of this PR so we can decide holistically how we want to handle this sort of situation. If we want to go this route (and I still think I could be convinced), I'm inclined to resurrect the principal ParamType and add the username resolution to it, but I think that will make this PR quite a bit larger.

Thoughts?

Reply:
I'm pretty clear on my long-term vision here. I'm not sure what we should do in the short term.

### Long Term

We already have a principal type in the CLI:
https://github.com/globus/globus-cli/blob/0791f166299a77ef2f00f889bf8ea88e26c6ffa6/src/globus_cli/parsing/param_types/identity_type.py#L36

It's just underpowered for what we need for Flows. And probably misnamed. And maybe the domain-handling case should be a special subclass. And... :sweat_smile: 

I'd like to fix this up into something which supports URN and non-URN syntax, can take options to enable or disable known special principals like `public`, etc, etc. All of that work is well beyond the scope of this PR.

There should be a type which handles all forms of Principals and which can be tuned and parametrized, or subclassed, as appropriate. We should gracefully handle errors from the services and have user-friendly support in the CLI, meaning that if the CLI is ever wrong, users experience a clear error.

I think this long term plan is somewhat uncontroversial -- it's a CLI which "does the right thing" in as many cases as possible.

Regarding the apparent conflict:
> this kind of contradicts the prior conversation we had as a group about wanting to have the validation of these values happen at the API-level so that it's controlled by the service.

I think my attitude is that a CLI which can do fancy things is very nice, and we should work on that. But a CLI which can't handle API errors correctly is broken and needs to be fixed.

I like the CLI to do minimal transformations to the data it receives, and trust service-side validation to work and to be rendered correctly. Drawing the line on "minimal transformations" is less science than art, but I think the error-handling behavior -- and the expectation that it's a rock-solid piece of the API-SDK-CLI interaction which we can rely upon -- is pretty straightforward.

### Short Term

We can drop this from the create flow command, but I think it needs to be a fast-follow.
URN-formatted principals are not particularly friendly as a CLI user experience, and I think users will find it difficult to use if we don't make accommodations.

For a user who isn't already writing data in the URN format, I think we need to look at what their experience will be if they want to share with a given user by username, e.g. `sirosen@globusid.org`. How do you convert that into an input for this command?

> Why, simply run with `--administrator "urn:globus:auth:identities:id:$(globus get-identities sirosen@globusid.org)"`!

That's not a user experience I want to advertise very much. It works, but it's not very easy or even very explainable.

Dropping any identity resolution from this PR might be the best choice, but I think it's important to solve this problem. Not so important that I'd object to doing a release without it though, so I'd be okay merging without it.

The main alternative I see is to define an interface here (no global / generic param type, just special-cased stuff for this command) which we want the long-term thing to conform to. That gives us tighter parameters on what the PrincipalType param type will have to support, but it also lets us more rapidly deliver a version of this command which supports our desired interface.

---

I guess I'm ping-pong-ing this back to you. How do you want to proceed? Intentionally leave it out, but agree that we want it handled in the near term? Or do a one-off here and let that guide how we want our generic solution to work?

Reply:
Per our discussion yesterday, I think we're in alignment on the eventual goal. I think that the thing we want to create for this is complex enough that it should be explicitly outside of this PR. I've created a story that attempts to sketch it out here: [[SC-24057](https://app.shortcut.com/globus/story/24057)].

I already have some initial work done toward this that I pulled out of this PR in an earlier phase and some general intuition about the shape this could take, and I'm happy to unstash into a new branch and continue that work once we've refined the story enough to be sure we all agree on what we want this to look like (such that it is of the greatest value possible for other CLI commands, both currently existing and in the future).

IMO, ideally this should support the functionality of the existing IdentityType enabling it to either replace IdentityType or allow IdentityType to become a lightweight subclass (I don't know if we want the b32 or domain stuff in the PrincipalType class itself...?).

Reply:
> (I don't know if we want the b32 or domain stuff in the PrincipalType class itself...?).

Yeah, this is a perfect example of "special weird extension", I think. Either it's a subclass or it's some opt-in behavior for the cases which want it.
I see the support for those being useful only for Transfer (where the b32-encoded IDs are used) and Auth (where you might want to cross-reference said data).

Let's consider this thread resolved. We're clear enough on what needs to happen that it shouldn't block this PR.